### PR TITLE
Sam tov new data generators

### PR DESCRIPTION
**Pull request by SamTov**
Added decision boundary data generators and did some small cleanup. Summary of changes:

* Added decision boundary class
    - Added circle and line discriminators
    - Added plot function to check data.
* Added tests for decision boundary generators
* Added tests for MNIST, CIFAR generators
* Added Abalone and MPG generators
* Added tests for Abalone and MPG generators
* Updated pre-commit versions and requirements.
* Fixed install issue in CI.

### CODE REVIEW

File: znnl/data/decision_boundary.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+from znnl.utils.prng import PRNGKey
+
+
+def linear_boundary(data: onp.ndarray, gradient: float, intercept: float) -> np.ndarray:
+    """
+    Create a linear boundary between classes.
+
+    Parameters
+    ----------
+    data : np.ndarray (n_samples, 2)
```
Comment (KonstiNik) [inline znnl/data/decision_boundary.py:43]: Types in the docstring do not match the type hints

File: znnl/data/decision_boundary.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                Number of samples to generate per class.
+        """
+        # Create the data-sets
+        data = onp.array(
+            jax.random.uniform(self.rng(), minval=0.0, maxval=1.0, shape=(n_samples, 2))
+        )
+        data = onp.clip(data, 0.0, 1.0)
+        targets = self.discriminator(data, *self.args)  # build classes (0, 1)
+
+        class_one_indices = np.where(targets == 0)[0]
```
Comment (KonstiNik) [inline znnl/data/decision_boundary.py:150]: Is there a reason to work with jax.numpy in the class and with onp in the functions above?

File: znnl/data/decision_boundary.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+import jax
+import jax.numpy as np
+import matplotlib.pyplot as plt
+import numpy as onp
+
+from znnl.data.data_generator import DataGenerator
+from znnl.utils.prng import PRNGKey
+
+
+def linear_boundary(data: onp.ndarray, gradient: float, intercept: float) -> np.ndarray:
```
Comment (KonstiNik) [inline znnl/data/decision_boundary.py:37]: This seems to be made for 2d data. 
Would it be more generic, to define a vector of n-dims instead of gradient along which the decision is made?

**Review by KonstiNik — COMMENTED** [review]
Looks nice. 
Just minor comments.

File: znnl/data/decision_boundary.py (hunk 3 above)
Comment (SamTov) [reply znnl/data/decision_boundary.py:37]: I guess you could do this for straight lines, but quickly becomes intractable for this case. We can think about it maybe for n dimensions.

File: znnl/data/decision_boundary.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    class 0 and outside are class 1.
+
+    Parameters
+    ----------
+    data : np.ndarray
+        Data to be converted into classes.
+    radius : float
+        Radius of the circle.
+    """
+    radii = onp.linalg.norm(data - 0.5, axis=1)
```
[the remark this answers is not in the document]
Comment (SamTov) [reply znnl/data/decision_boundary.py:75]: This centers the data into a box with dimensions 0, 1 so that I don't need negative numbers. The data is still randomly generated. This is just a coordinate transformation.

File: znnl/data/decision_boundary.py (hunk 3 above)
Comment (KonstiNik) [reply znnl/data/decision_boundary.py:37]: I thought you placed the two functions outside the class to generically make use of them. 
One could define a normal vector to the plane of separation, which would be generic for n-d. 
What do you think about that?

File: znnl/data/decision_boundary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KonstiNik) [reply znnl/data/decision_boundary.py:75]: I just meant that this implies the data to be generated in [0, 1]. 
As this function exists outside of the class I would just like to have the information be written in the docs.

File: znnl/data/decision_boundary.py (hunk 2 above)
Comment (KonstiNik) [reply znnl/data/decision_boundary.py:150]: What is the problem to identify it?

Comment (SamTov) [reply znnl/data/decision_boundary.py:152]: What do you mean

File: znnl/data/decision_boundary.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (SamTov) [reply znnl/data/decision_boundary.py:75]: The random data is generated inside a box of 0, 1 but this is done earlier in the random number generation call which is not accessible to the user. But I can document that.

File: znnl/data/decision_boundary.py (hunk 3 above)
Comment (SamTov) [reply znnl/data/decision_boundary.py:37]: Maybe we can discuss it in person

Comment (SamTov) [reply znnl/data/decision_boundary.py:37]: Opened #103 to address this in the future.

**Review by KonstiNik — APPROVED** [review]
Looks nice now.
